### PR TITLE
Mmap files

### PR DESCRIPTION
**Pull request by JanuszL**
- [x] test MAP_POPULATE
- [x] test memory consumption
- [x] fix problem when get_max_vm_cnt returns 0 for RecordIO
- [x] on one machine get_max_vm_cnt  returns 0 instead of sane number

### CODE REVIEW

File: dali/pipeline/operators/reader/loader/file_loader.h [hunk 1]
```diff
@@ -84,7 +84,7 @@ class FileLoader : public Loader<CPUBackend, ImageLabelWrapper> {
   }
 
   void PrepareEmpty(ImageLabelWrapper *tesor) override;
```
Comment (Kh4L) [inline dali/pipeline/operators/reader/loader/file_loader.h]:
```suggestion
  void PrepareEmpty(ImageLabelWrapper *tensor) override;
```

File: dali/util/local_file.cc [hunk 2]
```diff
@@ -15,9 +15,19 @@
...
+#include <unistd.h>
 #include <string>
 #include <cstdio>
 #include <cstring>
+#include <map>
+#include <mutex>
+#include <tuple>
 
 #ifdef _MSC_VER
 #include <Windows.h>
```
Comment (Kh4L) [inline dali/util/local_file.cc]: Do we plan to support Windows anytime soon?

File: dali/util/local_file.cc [hunk 3]
```diff
@@ -27,42 +37,160 @@
...
+  }
+  if (fd > -1) {
+    close(fd);
+  }
+  return p;
+}
+
 namespace dali {
 
+typedef std::tuple<void*, size_t> mapped_file;
```
Comment (Kh4L) [inline dali/util/local_file.cc]: Is there a specific reason why we are using std::tuple rather than std::pair here?

File: dali/util/local_file.cc [hunk 4]
```diff
@@ -27,42 +37,160 @@
...
+  if ((fd = open(path, O_RDONLY)) < 0) {
+    goto fail;
+  }
+
+  if (fstat(fd, &s) < 0) {
+    goto fail;
+  }
+
+  *length = (size_t)s.st_size;
+  if ((p = mmap(nullptr, *length, PROT_READ, MAP_PRIVATE, fd, 0)) == MAP_FAILED) {
```
Comment (Kh4L) [inline dali/util/local_file.cc]: Would `MAP_POPULATE` be of any use here?

File: dali/util/local_file.cc (hunk 2 above)
Comment (JanuszL) [reply dali/util/local_file.cc]: It is legacy. I will remove it.

File: dali/util/local_file.cc (hunk 3 above)
Comment (JanuszL) [reply dali/util/local_file.cc]: Legacy. I can change.

File: dali/util/local_file.cc (hunk 4 above)
Comment (JanuszL) [reply dali/util/local_file.cc]: I will try and check the perf. For small files it could work but it may be slow if we mmap 50GB.

File: dali/util/local_file.cc (hunk 3 above)
Comment (JanuszL) [reply dali/util/local_file.cc]: ... it seams that std::pair hits the perf (I don't know why but it is). I will leave it unchanged now.

File: docs/examples/pytorch/resnet50/main.py [hunk 5]
```diff
@@ -13,6 +13,7 @@
 import torch.utils.data
```
Comment (Kh4L) [inline docs/examples/pytorch/resnet50/main.py]: This would deserve a separate PR

Comment (JanuszL) [reply docs/examples/pytorch/resnet50/main.py]: This file is just for testing, I will remove it from PR.

File: dali/util/local_file.cc (hunk 4 above)
Comment (JanuszL) [reply dali/util/local_file.cc]: I tested if, perf is the same while startup is longer.

**Comment by JanuszL** [comment]
Builds 604379

**Comment by JanuszL** [comment]
With lint fixes - 604404

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h [hunk 6]
```diff
@@ -49,13 +49,27 @@ class IndexedFileLoader : public Loader<CPUBackend, Tensor<CPUBackend>> {
       current_file_ = FileStream::Open(uris_[file_index]);
       current_file_index_ = file_index;
     }
-    tensor->Resize({size});
-    tensor->mutable_data<uint8_t>();
 
-    int64 n_read = current_file_->Read(reinterpret_cast<uint8_t*>(tensor->raw_mutable_data()),
-                        size);
+    if (copy_read_data_) {
```
Comment (mzient) [inline dali/pipeline/operators/reader/loader/indexed_file_loader.h]: Shouldn't it be `!copy_read_data_` here?

File: dali/util/local_file.cc [hunk 7]
```diff
@@ -15,54 +15,180 @@
...
+    p = nullptr;
+    goto fail;
+  }
+  // This doesn't seem to have a real impact on performance, but theoretically it can help.
+  madvise(p, *length, MADV_HUGEPAGE);
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/util/local_file.cc]: https://stackoverflow.com/questions/6575010/ubuntu-10-04-error-in-using-map-hugetlb-with-map-shared - it doesn't have much sense to use MAP_HUGETLB for mmaped files. So removing.

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h (hunk 6 above)
Comment (JanuszL) [reply dali/pipeline/operators/reader/loader/indexed_file_loader.h]: You are right...

**Comment by JanuszL** [comment]
Rebuilds 604985

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h (hunk 6 above)
Comment (JanuszL) [reply dali/pipeline/operators/reader/loader/indexed_file_loader.h]: Done

File: dali/util/local_file.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/util/local_file.cc]: And more info https://lwn.net/Articles/718102/ and https://stackoverflow.com/questions/44060678/huge-pages-for-memory-mapped-files-on-linux

File: dali/util/local_file.cc [hunk 8]
```diff
@@ -15,54 +15,178 @@
...
+  if ((fd = open(path, O_RDONLY)) < 0) {
+    goto fail;
+  }
+
+  if (fstat(fd, &s) < 0) {
+    goto fail;
+  }
+
+  *length = (size_t)s.st_size;
+  if ((p = mmap(nullptr, *length, PROT_READ, MAP_PRIVATE, fd, 0)) == MAP_FAILED) {
```
Comment (mzient) [inline dali/util/local_file.cc]: For large files, you might want to try MAP_HUGETLB flag (should work better than madvise, since mmap would make these pages huge from the very beginning).

File: dali/util/local_file.cc (hunk 3 above)
Comment (Kh4L) [reply dali/util/local_file.cc]: Why would this hit the perf though?

File: dali/pipeline/operators/reader/loader/recordio_loader.h [hunk 9]
```diff
@@ -19,11 +19,14 @@
 #include <memory>
 #include <string>
 #include <vector>
+#include <chrono>
 
 #include "dali/pipeline/operators/reader/loader/indexed_file_loader.h"
 #include "dali/common.h"
 #include "dali/error_handling.h"
 
+typedef std::chrono::high_resolution_clock Clock;
```
Comment (Kh4L) [inline dali/pipeline/operators/reader/loader/recordio_loader.h]: It does not seem to be used anywhere

File: dali/util/local_file.cc [hunk 10]
```diff
@@ -15,54 +15,178 @@
...
+  return p;
+}
+
 namespace dali {
 
+typedef std::tuple<void*, size_t> mapped_file;
+typedef std::tuple<std::string, size_t> mapped_file_rev;
+
+// limit to half of allowed mmaped files
+static const unsigned int DALI_MAX_MV_CNT = get_max_vm_cnt() / 2;
```
Comment (mzient) [inline dali/util/local_file.cc]: UPPERCASE_IDENTIFIER suggests a compile-time constant or even a `#define`, not a _variable_ (albeit write-once).

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h [hunk 11]
```diff
@@ -49,13 +49,27 @@ class IndexedFileLoader : public Loader<CPUBackend, Tensor<CPUBackend>> {
...
+
+      TypeInfo type;
+      type.SetType<uint8_t>();
+      tensor->set_type(type);
+    } else {
+      tensor->Resize({size});
+      tensor->mutable_data<uint8_t>();
+
+      int64 n_read = current_file_->Read(reinterpret_cast<uint8_t*>(tensor->raw_mutable_data()),
+                          size);
```
Comment (jantonguirao) [inline dali/pipeline/operators/reader/loader/indexed_file_loader.h]: the indentation seems a bit random in this line. Are we following any rules regarding that?

File: dali/util/local_file.cc (hunk 3 above)
Comment (jantonguirao) [reply dali/util/local_file.cc]:
```suggestion
using MappedFile = std::tuple<void*, size_t>;
```

File: dali/pipeline/data/tensor.h [hunk 12]
```diff
@@ -218,6 +218,37 @@ class Tensor : public Buffer<Backend> {
...
+   * of shape vector, and its type is reset to NoType. Future calls to
+   * Resize or setting of the Tensor type will evaluate whether or not the
+   * current allocation is large enough to be used and proceed appropriately.
+   *
+   * The Tensor object assumes no ownership of the input allocation, and will
+   * not de-allocate it when it is done using it. It is up to the user to
+   * manage the lifetime of the allocation such that it persist while it is
+   * in use by the Tensor.
+   */
+  inline void ShareData(void *ptr, size_t bytes, const vector<Index> &shape) {
```
Comment (jantonguirao) [inline dali/pipeline/data/tensor.h]: This code seems a duplicated from the overload in line 206. Can you write one version with a default argument 'shape = {}' ?

File: dali/pipeline/data/tensor.h [hunk 13]
```diff
@@ -218,6 +218,37 @@ class Tensor : public Buffer<Backend> {
...
+    DALI_ENFORCE(ptr != nullptr, "Input pointer must not be nullptr.");
+
+    // Save our new pointer and bytes. Reset our type, shape, and size
+    data_.reset(ptr, [](void *) {});
```
Comment (jantonguirao) [inline dali/pipeline/data/tensor.h]: curious, why do we need a smart pointer if we don't use the deleter?

File: dali/util/local_file.cc [hunk 14]
```diff
@@ -15,54 +15,178 @@
...
+      size_t length;
+      std::tie(file_name, length) = it->second;
+      mapped_files.erase(file_name);
+      mapped_files_rev.erase(p);
+      munmap(p, length);
+    }
+  }
+}
+
+void LocalFileStream::Put(const std::string& path) {
```
Comment (jantonguirao) [inline dali/util/local_file.cc]: duplicated code in this overload. Can we unify?

**Comment by JanuszL** [comment]
@mzient 
>  have a few general comments to this change:
>I don't like the idea of "caching the lucky first N" files and then resorting to reading.

We are not - we checking ahead if we are able to cache all files that are in use or we need to go with reading. And alleyways we do mmap, the question is is we can share the mmaped memory or we need to copy it. Only exception is RecordIO which can have chunk of data spread across more than one file.

> It introduces some global state and can be driven into irreparable state of running out of mapping slots. Is calling mmap multiple times that expensive?

We want to save syscalls and that helps.

>mmap with MAP_HUGETLB may be cheaper due to the smaller number of pages involved. madvise is unlikely to change anything after mmap has mapped the into memory that is not likely to be aligned on a 2MB boundary (huge page size in x86_64).

I can force 2MB alignment as a forward looking feature.

>EDIT: I've read your post on HUGETLB - it seems like it's onle being added at the moment - if it's available at all, it's only in latest kernels. Our "manylinux" compatibility may suffer.

manylinux has very little to do with that - it is OS kernel that runs DALI, not the system you are using to build it that provides HUGETLB for files. 

> Idea about resource management:
We could have a custom deleter that itself holds a shared_ptr to a mapped file object. The mapping cache would also use shared_ptr. If a call to mmap fails (or some common sense number of mapped files is exceeded), we may apply LRU strategy to unused mappings in the cache. If nothing could be freed, only then we'd have to read (instead of mapping).

As we are providing memory to the tensor and share it with it we will need to keep the shared_ptr in the tensor as well. I don't know if this is a good idea to imply that kind of dependency and make Tensor aware about shared_ptr to the mmaped memory.

File: dali/pipeline/data/tensor.h (hunk 13 above)
Comment (JanuszL) [reply dali/pipeline/data/tensor.h]: We use - please look into ResizeHelper.

File: dali/pipeline/data/tensor.h (hunk 12 above)
Comment (JanuszL) [reply dali/pipeline/data/tensor.h]: Will do

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h (hunk 11 above)
Comment (JanuszL) [reply dali/pipeline/operators/reader/loader/indexed_file_loader.h]: Will do

File: dali/util/local_file.cc (hunk 8 above)
Comment (JanuszL) [reply dali/util/local_file.cc]: Passing MAP_HUGETLB makes mmap to fail, so we cannot do that now. I can force memory alignment and use madvise.

File: dali/pipeline/data/tensor.h (hunk 12 above)
Comment (JanuszL) [reply dali/pipeline/data/tensor.h]: Done

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h (hunk 11 above)
Comment (JanuszL) [reply dali/pipeline/operators/reader/loader/indexed_file_loader.h]: Done (I hope)

File: dali/util/local_file.cc (hunk 8 above)
Comment (JanuszL) [reply dali/util/local_file.cc]: Ehh, we cannot force address aliment. We can provide desired address by this is only a hint and system (Linux) will put memory aligned to the next available page.

**Comment by JanuszL** [comment]
Build 607487

**Comment by JanuszL** [comment]
Rebuilds 607568

File: dali/util/local_file.cc (hunk 3 above)
Comment (klecki) [reply dali/util/local_file.cc]: @mzient AFAIK storage order of tuple is implementation defined.

Comment (JanuszL) [reply dali/util/local_file.cc]: @Kh4L - it could be a run2run variance, but it is easier to extend having tuple than be bound to the std::pair API.

File: dali/util/local_file.cc (hunk 8 above)
Comment (mzient) [reply dali/util/local_file.cc]: OK - we can just ignore it then, I guess...

**Comment by JanuszL** [comment]
Builds after rebase.

**Comment by JanuszL** [comment]
Build with lint fixes 612521
